FLT/AutomorphicForm/QuaternionAlgebra/FiniteDimensional.lean

```diff
@@ -1,31 +1,124 @@
 for `Dˣ` is finite-dimensional over `K`.
 -/
-theorem WeightTwoAutomorphicForm.finiteDimensional :
-    FiniteDimensional K (WeightTwoAutomorphicFormOfLevel U K) :=
-  sorry
+theorem WeightTwoAutomorphicForm.finiteDimensional
+    {hU : IsOpen (U : Set (Dfx (F:=F) (D:=D)))} :
+    FiniteDimensional K (WeightTwoAutomorphicFormOfLevel U K) := by
+    let U' : Subgroup (Dfx F D) := U
+    let H' : Subgroup (Dfx F D) := {
+        carrier := Set.range (incl₁ F D),
+        one_mem' := ⟨1, by simp⟩,
+        mul_mem' := by
+            rintro _ _ ⟨a, rfl⟩ ⟨b, rfl⟩
+            exact ⟨a * b, by simp⟩,
+        inv_mem' := by
+            rintro _ ⟨a, rfl⟩
+            exact ⟨a⁻¹, by simp⟩
+    }
+    -- We define a free K-module W with a basis indexed by
+    -- the elements of a double coset space which (in the totally
+    -- definite case) is finite)
+    let X := Doset.Quotient (Set.range (incl₁ F D)) U'
+    have h : Finite X :=
+        NumberField.FiniteAdeleRing.DivisionAlgebra.finiteDoubleCoset (K:=F) (D:=D) hU
+
+    letI := Fintype.ofFinite X
+    let m := Fintype.card X
+    let W := Fin m → K
+    let V := (WeightTwoAutomorphicFormOfLevel U K)
+
+    let index := (Fintype.equivFin X : X ≃ Fin m)
+    let rep : X → Dfx F D := Quot.out
+
+    -- We then define a linear map φ from V to the finite-dimensional space W.
+    -- V is a space of functions, and the map consists of evaluating
+    -- a function on representatives given by the rep function above.
+    let φ : V →ₗ[K] W :=
+        {
+        toFun := fun v i ↦ v.1.toFun (rep (index.symm i)),
+        map_add' := by
+            intros v₁ v₂
+            ext i
+            rfl
+        map_smul' := by
+            intros c v
+            ext i
+            rfl
+        }
+    -- Since we have a linear map φ from V to W and W is finite-dimensional,
+    -- it's enough to check that φ is injective.
+    have φ_injective : Function.Injective φ := by
+        intro v₁ v₂ h
+        let f₁ : WeightTwoAutomorphicForm F D K := v₁.1
+        let f₂ : WeightTwoAutomorphicForm F D K := v₂.1
+        -- Show v₁.toFun = v₂.toFun because they agree on reps and the
+        -- space is determined by those values
+        have fun_eq : ∀ d, f₁.toFun d = f₂.toFun d := by
+            intro d
+            let d' := rep (Quot.mk _ d)
+            -- Because d' is a representative for the double coset containing d
+            have hcoset : ∃ γ : Dˣ, ∃ u ∈ U, d = (incl₁ F D γ) * d' * u := by
+                have h_rel : (Doset.setoid H' U') d' d := by
+                  exact Quotient.exact (Quotient.out_eq ⟦d⟧)
+                -- Apply Doset.rel_iff to extract the witnesses
+                rw [Doset.rel_iff] at h_rel
+                obtain ⟨h, hh, k, hk, h_eq⟩ := h_rel
+                obtain ⟨γ, hγ⟩ := hh
+                use γ, k, hk
+                rw [← hγ] at h_eq
+                exact h_eq
+            obtain ⟨γ, u, hu, hd⟩ := hcoset
+            have h1 : f₁.toFun d = f₁.toFun d' := by
```

Comment:
Yeah this isn't what to prove. You should be proving that for all weight 2 automorphic forms, they have this property (and then apply it twice), and my instinct is that this should be a separate lemma rather than being proved in the middle of this lemma (and certainly rather than being proved twice in the middle of this lemma!). It's always good to factor out sublemmas.